src/Dialect/ONNX/MLIRDialectBuilder.hpp

```diff
@@ -166,21 +166,28 @@ struct VectorBuilder final : DialectBuilder {
   VectorBuilder(OpBuilder &b, Location loc) : DialectBuilder(b, loc) {}
   VectorBuilder(DialectBuilder &db) : DialectBuilder(db) {}
 
+  // Get the machine SIMD vector length for the given elementary type.
+  // This can help guide certain optimizations.
+  int64_t getMachineVectorLength(const Type &elementType);
+  int64_t getMachineVectorLength(const VectorType &vecType);
+  int64_t getMachineVectorLength(Value vecValue);
+
   Value load(VectorType vecType, Value memref, ValueRange indices = {}) const;
   void store(Value val, Value memref, ValueRange indices = {}) const;
 
   Value broadcast(VectorType vecType, Value val) const;
   Value shuffle(Value lhs, Value rhs, SmallVectorImpl<int64_t> &mask) const;
   Value fma(Value lhs, Value rhs, Value acc) const;
 
-  // Composite functions
-  Value mergeLow(Value lhs, Value rhs, int64_t step);
+  // Composite functions.
   Value mergeHigh(Value lhs, Value rhs, int64_t step);
-  Value multiReduction(SmallVectorImpl<Value> &vecArray); // Only 4x4 as of now.
+  Value mergeLow(Value lhs, Value rhs, int64_t step);
+  void multiReduction(SmallVectorImpl<Value> &inputVecArray,
+      SmallVectorImpl<Value> &outputVecArray);
```

Comment:
Consider making these `const` too.

Reply:
Made all of them const except the multiReduction because it internally create a `MathBuilder` and currently the constructor does not use a const on the constructor's input.